[V-3] Refactor email verification system to C++

KyryloKupin opened the pull request:
The old system files `vrf.h` and `vrf.c` have been removed and replaced with a C++ version `EmailVerification.hpp` and `EmailVerification.cpp`. Additionally, several adjustments have been made on the CMakeLists.txt to properly build the new C++ source files.

KyryloKupin committed:
Refactor email verification system to C++

This commit updates the email verification system from C to C++. The old system files `vrf.h` and `vrf.c` have been removed and replaced with a C++ version `EmailVerification.hpp` and `EmailVerification.cpp`. Additionally, several adjustments have been made on the CMakeLists.txt to properly build the new C++ source files.

KyryloKupin committed:
Bring back vrf.c

KyryloKupin committed:
Remove vrf.c

irog-c commented on include/EmailVerification.hpp:
```diff
@@ -0,0 +1,80 @@
+        bool catchAll;
+    };
+
+    auto extractLocalPartAndDomain(const std::string &email) -> std::tuple<std::string, std::string>;
```
Why `std;:tuple` why not `std::pair`?

KyryloKupin replied:
Good point!

irog-c commented on src/EmailVerification.cpp:
```diff
@@ -0,0 +1,293 @@
+    }
+
+    return "doesn't";
+}
```
Why not an enum here? Or is it because we return it to the formatted string and this is easier?

KyryloKupin replied:
This function isn't important now

irog-c commented on src/EmailVerification.cpp:
```diff
@@ -0,0 +1,293 @@
+    return "doesn't";
+}
+
+auto printVerificationData(std::ostream &os, EmailVerificationData emailVerificationData) {
```
Why not return `std::ostream&` here?

irog-c commented on src/EmailVerification.cpp:
```diff
@@ -0,0 +1,293 @@
+        std::cerr << e.what() << std::endl;
+    }
+
+#ifdef _WIN32
```
Is there any specific reason for why we would want Windows support? In the long run it just seems like a big burden to support both Linux and Windows? What do you think?

KyryloKupin replied:
I think we should switch to boost::asio eventually, then it won't be necessary. I am on Windows right now, so it's easier this way.

irog-c commented on include/EmailVerification.hpp:
```diff
@@ -0,0 +1,80 @@
+#include "config.h"
+
+namespace TupoSoft::VRF {
+    typedef enum
```
Why `typedef enum`? Underneath this you used `enum class`.

KyryloKupin replied:
Because the refactoring isn't complete yet

KyryloKupin replied:
We should get rid of it eventually, since we have exception in C++

KyryloKupin committed:
Update CMake configuration and refactor email verification code

The CMake configuration has been significantly updated to enforce in-source builds, include package dependencies using CPM.cmake and structure the project for find_package() compatibility. The email verification code has been modified, with sections commented out and different return types for some functions. Test files have also been added.

KyryloKupin committed:
Remove unused macros and functions from EmailVerification.cpp

The commit removes redundant macros and functions from 'EmailVerification.cpp' that were not being used. This includes the 'FREE(p)' and 'CHECK_OK(f, err)' macros, and 'email_exists' function. These changes make the code cleaner and more maintainable by reducing unnecessary clutter.

KyryloKupin committed:
Refactor EmailVerification.hpp file

Removed unnecessary "config.h" include and VRF_err_t enum in EmailVerification.hpp. This simplifies the code, streamlines dependencies and improves readability.

KyryloKupin committed:
Refactor CMake build system and update EmailVerification.cpp

Introduced a centralized CMake configuration (project.cmake) that handles the CMake requirements and project definitions, enabling more consistent configuration across main and test projects. The CMake configuration changes in CMakeLists.txt and test/CMakeLists.txt reflect this update. Removed unnecessary code from the EmailVerification.cpp and fixed variable naming conventions. Lastly, added a basic Google Test setup in the test project.

KyryloKupin committed:
Update build scripts and refactor email verification function

This commit includes the necessary changes in the build scripts for Windows platform compatibility by linking ws2_32 library. The test build script is also refactored to use variable BASE_PROJECT_NAME instead of hard coded name, easing the project renaming process. Lastly, it simplifies the email verification function in EmailVerification.cpp, getting rid of redundant code, and improves error handling. Corresponding tests are also updated to adjust to these changes.

irog-c commented on source/EmailVerification.cpp:
```diff
@@ -0,0 +1,141 @@
+
+    auto mxRecordInfoList = new addrinfo;
+    if (const auto error = getaddrinfo(domain.c_str(), SMTP_SERVICE, &addrinfoHints, &mxRecordInfoList)) {
+        throw std::runtime_error(std::format("Failde to get address info for domain {}, error: {}", domain,
```
nit: typo
Should be `Failed` instead of `Failde`

irog-c commented on source/EmailVerification.cpp:
```diff
@@ -0,0 +1,141 @@
+    const std::unique_ptr<addrinfo, decltype(&freeaddrinfo)> mxRecordInfos{mxRecordInfoList, freeaddrinfo};
+
+    std::vector<std::string> recordAddresses{};
+    for (const addrinfo *mxRecordInfo = mxRecordInfos.get(); mxRecordInfo != nullptr;
```
Why not use `auto` here?

KyryloKupin replied:
This function doesn't do the job of getting mx records anyway :)

irog-c commented on source/EmailVerification.cpp:
```diff
@@ -0,0 +1,141 @@
+    std::vector<std::string> recordAddresses{};
+    for (const addrinfo *mxRecordInfo = mxRecordInfos.get(); mxRecordInfo != nullptr;
+         mxRecordInfo = mxRecordInfo->ai_next) {
+        char address[INET6_ADDRSTRLEN];
```
You can probably use a `std::array` or `std::vector` instead of a C-style array.

KyryloKupin replied:
std::string might work as well, right?

irog-c replied:
I don't think `std::string` should be used like this, in the case where we want to copy the `std::array`'s contents into a string, we can use a `std::back_inserter` like so:
```cpp
   auto arr = std::array<char, 16> {
        't',
        'e',
        's',
        't'
    };
    auto str = std::string();
    std::copy(arr.cbegin(), arr.cend(), std::back_inserter(str));
```

irog-c commented on test/source/EmailVerification.cpp:
```diff
@@ -0,0 +1,16 @@
+
+using namespace TupoSoft::VRF;
+
+TEST(EmailVerificationTest, ExtractLocalPartAndDomain) {
```
We can add a test to see if an error/exception is properly returned in case of bad input?

KyryloKupin replied:
Done

KyryloKupin committed:
Refactor EmailVerification.cpp code for better readability

This commit addresses some readability improvements in EmailVerification.cpp. It corrects a typo in the error message and refactors the way the address info for each MX record is obtained by switching from a char array to a std::string, making the code more consistent with modern C++ standards.

KyryloKupin committed:
Refactor email extraction code for simplicity

The code in EmailVerification.cpp that extracts the local part and domain from an email has been refactored to make it more readable and concise. The change primarily involves replacing explicit types with auto and using list initialization for pair return.

KyryloKupin committed:
Update email verification functions and tests

The return type of the extractLocalPartAndDomain function in the EmailVerification file has been clarified. Also, new test case 'ThrowsInvalidArgumentErrorOnInvalidInput' has been added to EmailVerificationTest while the existing test has been renamed to 'ExtractLocalPartAndDomainSuccess'.

KyryloKupin committed:
Refactor EmailVerificationTest for simplicity

The email verification test for invalid inputs has been refactored for simplicity. Instead of defining a variable for the invalid email, the invalid email string has been put directly into the 'extractLocalPartAndDomain' function.

KyryloKupin committed:
Update Windows DNS query method and link additional library

Updated the method of querying DNS on Windows platform to use DnsQuery_A function from dnsapi library, leading to restructuring and improving some related code. Introduced the additional linking of "dnsapi" library on Windows for using DnsQuery_A function. Enhanced test coverage for MX record fetching mechanism.

KyryloKupin committed:
Refactor EmailVerification.cpp to remove unused variable

The verbose variable, which was declared but not used, has been removed from the EmailVerification.cpp file. For improved initialization, the pDnsRecord variable is now initialized with a default value.

KyryloKupin committed:
Refactor email verification code

The changes involve moving the '#include <windns.h>' directive from 'EmailVerification.hpp' to 'EmailVerification.cpp'. Additionally, unnecessary includes, such as 'iostream', 'cstdio', and 'getopt.h', have been removed from 'EmailVerification.hpp', as well as the 'MxQueryResult' structure. Header file order has also been adjusted in 'EmailVerification.cpp'. These changes simplify the code and improve readability.

KyryloKupin committed:
Update test case in EmailVerification.cpp

The test case 'GetMXRecordsSuccess' in EmailVerification.cpp has been updated. The "domain" constant now uses auto type deduction, and the "result" constant is now generated by appending the domain to 'mail.'. This ensures the test case is more dynamic and accurate.

KyryloKupin committed:
Remove redundant test for MX records retrieval

A test in the EmailVerification test suite was removed because it was redundant. The test, GetMXRecordsDoesNotThrow, was testing a scenario already covered by another test, GetMXRecordsSuccess. The redundancy was removed to simplify the test suite.

KyryloKupin committed:
Refactor DNS query error handling in EmailVerification

The DNS query error handling in the 'getMXRecords' method of the EmailVerification class has been simplified. The DNS query is now only performed once, and the result is stored in a variable. This results in cleaner, more efficient code.

KyryloKupin committed:
Add cross-platform support for fetching MX records

The code has been adjusted to support DNS MX record retrieving on both Windows and Unix-based platforms. The appropriate dependencies have been added in CMakeLists.txt. In EmailVerification.cpp, OS-specific code has been put into separate areas using preprocessor directives, enabling execution on both platforms.

KyryloKupin committed:
Refactor variable initialization in EmailVerification

Removed unnecessary initialization of 'len' variable and replaced int type counter with auto keyword in a loop within EmailVerification.cpp for code optimization.

irog-c commented on source/EmailVerification.cpp:
```diff
@@ -134,7 +134,7 @@ auto TupoSoft::VRF::getMXRecords(const std::string &domain) -> std::vector<std::
-    int len{};
+    int len;
 
     const std::unique_ptr<struct __res_state, decltype(&res_nclose)> resStatePtr(new struct __res_state, res_nclose);
```
You don't need `struct` here, C++ supports just the type name without the `struct` specifier.

KyryloKupin replied:
It didn't work without `struct`, you can try to compile without it

irog-c replied:
In that case, you can probably use `std::make_unique`, at the very least it will simplify the lvalue type to `auto` also you don't have to use `new`.

KyryloKupin replied:
There is a custom deleter for this C structure, that's why std::make_unique can't be used here

irog-c replied:
You can do it by using the keyword `using` and defining a custom `unique_ptr` type, as explained [here](https://www.reddit.com/r/cpp_questions/comments/19385sw/how_to_create_a_unique_ptr_with_custom_deleter/), but either way it seems like a hassle to do so it is better to keep it like it is.

KyryloKupin replied:
This code is temporary anyway. We need a library for DNS querying.

irog-c commented on source/EmailVerification.cpp:
```diff
@@ -0,0 +1,172 @@
+
+    if (const auto status = DnsQuery_A(domain.c_str(), DNS_TYPE_MX, DNS_QUERY_STANDARD, nullptr, &pDnsRecord,
+                                       nullptr)) {
+        throw std::runtime_error("DNS query failed with error code: " + std::to_string(status));
```
You can probably use `std::format` here.

irog-c commented on test/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+
+# ---- Options ----
+
+option(ENABLE_TEST_COVERAGE "Enable test coverage" OFF)
```
What's the point of this option if we never do anything with it and always do `enable_testing()`?

KyryloKupin replied:
It's for codecov, not set up yet

irog-c commented on source/EmailVerification.cpp:
```diff
@@ -0,0 +1,172 @@
+
+    for (auto ns_index = 0; ns_index < ns_msg_count(handle, ns_s_an); ns_index++) {
+        if (ns_parserr(&handle, ns_s_an, ns_index, &rr)) {
+            continue;
```
`continue` should be avoided, refer to [this](https://rules.sonarsource.com/cpp/tag/bad-practice/RSPEC-909/).

KyryloKupin committed:
Refactor error handling in EmailVerification

This commit refactors error handling in the EmailVerification module. CPP's `std::format` has been utilized to unify error message formatting. Additionally, a conditional statement within the DNS parsing loop was simplified for better readability and efficiency.

KyryloKupin committed:
Add test coverage options to CMake

The CMakeLists.txt file has been updated to include test coverage options. These options are only applied when the ENABLE_TEST_COVERAGE flag is set. They allow for better tracking and optimization of code test coverage.

KyryloKupin committed:
Add Ubuntu and Windows workflows to Github Actions

Created two new Github Action workflows, ubuntu.yml and windows.yml. Both workflows are triggered by push or pull request events to the master and main branches, performing build, configure and test jobs on Ubuntu and Windows respectively. Ubuntu workflow includes an additional step for code coverage report collection.

KyryloKupin committed:
Update Github actions workflow for ubuntu

The Github actions workflow for ubuntu has been updated with new steps. The 'actions/checkout' and 'actions/cache' tools have been upgraded to Version 4. Additionally, the workflow now also includes steps for setting up Node.js, providing npm help, installing the g++13 compiler, and using it to get its version.

KyryloKupin committed:
Update ubuntu version and remove compiler setup in workflow

This commit updates the version of Ubuntu in the Github Actions from 'ubuntu-latest' to 'ubuntu-24.04'. Additionally, it removes the steps for setting up the C++ compiler 'g++13', since it is no longer necessary. This simplifies the workflow and ensures better compatibility with the updated Ubuntu version.

KyryloKupin committed:
Enable shared CRT for MSVC in gtest and include iostream in EmailVerification

Shared CRT for MSVC is now enabled in googletest to mitigate issues potentially arising from having different CRTs in different binaries. The iostream library, on the other hand, was needed in EmailVerification.cpp to support input/output operations.

KyryloKupin committed:
Update Ubuntu version and add fmtlib as dependency

The Ubuntu version in the Github workflow has been updated to the latest version. Additionally, the fmtlib library has been added to the project as a dependency, which required changes in the CMakeLists.txt file and in the EmailVerification.cpp file to accommodate the new format library.

irog-c commented on .github/workflows/windows.yml:
```diff
@@ -0,0 +1,38 @@
+  pull_request:
+    branches:
+      - master
+      - main
```
Why have both `master` and `main` branches here?

KyryloKupin committed:
Disable Windows Github workflow

Renamed the windows.yml file to windows.yml.disabled effectively disabling the Windows Github workflow. This is part of temporary measures to troubleshoot build issues.

KyryloKupin committed:
Remove main branch and coverage collection from ubuntu workflow

The main branch has been removed from the push and pull request triggers in the ubuntu.yml workflow. Also, the collection of code coverage reports and the ENABLE_TEST_COVERAGE flag during configuration have been removed.

KyryloKupin committed:
Refactor printVerificationData to operator overload

The printVerificationData function was moved into the TupoSoft::VRF namespace and refactored, changing from a free function to an operator overload. This change provides a more intuitive way to print EmailVerificationData, and also better aligns with object-oriented design principles. Tests are updated to validate this behavior.

KyryloKupin committed:
Updated terminology in EmailVerification output

The terminology used in the EmailVerification output and test has been updated for clarity. The "local part" of the email address is now referred to as "username", and the term "mx record" has been changed to "mx_record" to maintain consistency in variable naming.

KyryloKupin committed:
Rename test case in EmailVerification

This commit renames the test case ExtractLocalPartAndDomainSuccess to ExtractUsernameAndDomainSuccess in the EmailVerification test source file. This update better represents the functionality being tested.

KyryloKupin committed:
Add .clang-format for C++ styling

A .clang-format file has been added to establish consistent coding style for C++ projects. This configuration is generated based on the LLVM style and includes specifications for things like brace wrapping, column limits, indentations, namespace rules, and spacing rules.

KyryloKupin committed:
Refactor EmailVerification namespace and variable naming

Namespace and variable naming in EmailVerification have been refactored. This includes changing the namespace from TupoSoft::VRF to tuposoft::vrf and standardizing variable names to use lower case and underscores. Refactor also extends to renaming functions and struct elements for greater comprehension and conformity with the broader codebase.

irog-c committed:
Use c-ares library for fetching mx records

KyryloKupin committed:
Add Boost library to CMakeLists

Boost library version 1.85.0 was added to the CMakeLists.txt. It was also added to the project dependencies, specifically Boost::asio submodule, providing necessary libraries for network and low-level I/O operations.

irog-c committed:
Merge branch 'v3-cpp' into add-c-ares-library

irog-c committed:
Merge pull request #4 from TupoSoft/add-c-ares-library

[V-3] Add `c-ares` library

irog-c committed:
Attempt to fix boost library failing to configure build tree

irog-c committed:
Second attempt at fixing

KyryloKupin committed:
Update dependencies and build configurations

Removed the _WIN32_WINNT definition in EmailVerification.hpp and redefined it in CMakeLists.txt. Boost ASIO has been replaced with standalone ASIO dependency. Configuration for build dependencies and inclusion of directories has been adjusted accordingly to support these changes.

KyryloKupin committed:
Remove unnecessary FetchContent inclusion

The FetchContent module was formerly included in the CMakeLists.txt file, but this inclusion was detected as redundant. This commit removes it to streamline the setup process and to avoid potential conflicts or misuse in future developments.

KyryloKupin committed:
Refactor code in EmailVerification.cpp

The code in EmailVerification.cpp has been updated. The initialization of the ARES library has been switched to use ares_init_options, and the variable declaration has been streamlined, using more compact syntax. The use of nullptr has also been introduced for cleaner code. Overall, these changes contribute to better readability and more efficient code.

KyryloKupin committed:
Refactor email verification functionality

The ARES library initialization in EmailVerification.cpp is now using `ares_init_options`, improving the function's efficiency. Other changes include cleaner variable declarations, and the introduction of `nullptr` for a more readable code structure. These updates significantly increase code readability and performance.

irog-c commented on source/EmailVerification.cpp:
```diff
@@ -0,0 +1,165 @@
+    return mx_records;
+}
+
+std::ostream &tuposoft::vrf::operator<<(std::ostream &os, const vrf_data &data) {
```
Instead of
```
std::ostream &tuposoft::vrf::operator<<(std::ostream &os, const vrf_data &data)
```
you can maybe do
```
auto tuposoft::vrf::operator<<(std::ostream &os, const vrf_data &data) -> decltype(os)
```

irog-c commented on source/EmailVerification.cpp:
```diff
@@ -0,0 +1,165 @@
+
+    // Initialize the library
+    if (ares_init_options(&channel, nullptr, 0) != ARES_SUCCESS) {
+        return mx_records;
```
Should we throw here?

KyryloKupin replied:
Probably, I've not yet addressed the unhappy scenario

irog-c commented on test/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+enable_testing()
+
+file(GLOB SOURCES CONFIGURE_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/source/*.cpp)
+add_executable(${PROJECT_NAME} ${SOURCES})
```
How do we add an executable if there is no `main` function? This is a library, we should probably separate this code and put it in a library, and have another file which can be called `main.cpp` which uses it, or something like that, what do you think?

KyryloKupin replied:
Correct, but the verification won't work because there are functions that we haven't finished rewriting from C.

irog-c commented on source/EmailVerification.cpp:
```diff
@@ -0,0 +1,165 @@
+            +[](void *arg, const ares_status_t status, size_t timeouts, const ares_dns_record_t *dnsrec) {
+                auto *mxrs = static_cast<std::vector<std::string> *>(arg);
+                if (status != ARES_SUCCESS) {
+                    return; // Handle error: status will tell you what went wrong
```
Should we handle the error here as the comment suggests?

KyryloKupin replied:
Yep, but not that important for now

KyryloKupin committed:
Update CMakeLists and EmailVerification header and source files

The CMakeLists files were updated to use WIN32 instead of MSVC and to reference the project namespace. Removed unused headers from EmailVerification.hpp and revised the structure of lambdas in source. Declared the operator<< return type explicitly. Removed dependencies in system-level CMakeLists.

KyryloKupin committed:
Update email verification code for better handling of MX records

Revamped the get_mx_records function in EmailVerification.cpp. The method now utilizes asynchronous multi-socket handling in c-ares for DNS records, providing improved functionality and reliability. Additionally, introduced structured data to manage file descriptor sets, boosting efficiency in socket handling.

KyryloKupin committed:
Update email verification code for better handling of MX records

Revamped the get_mx_records function in EmailVerification.cpp. The method now utilizes asynchronous multi-socket handling in c-ares for DNS records, providing improved functionality and reliability. Additionally, introduced structured data to manage file descriptor sets, boosting efficiency in socket handling.

KyryloKupin committed:
Remove unused c-ares header from EmailVerification.cpp

The c-ares header is no longer needed in the EmailVerification component. This commit cleans up unnecessary dependencies, thereby keeping the codebase cleaner and minimizing potential for confusion or errors.

KyryloKupin committed:
Remove retry limit in email verification

The retry limit in the email verification process was removed. This change was done to increase the reliability of the email verification feature, removing the .tries option and its related flag from the ares_init_options function. Now, it will keep trying until a successful response is received.

KyryloKupin committed:
Remove unused thread library in EmailVerification

The <thread> library inclusion in EmailVerification.cpp was removed since it was not being used. This reduces unnecessary dependencies, promoting cleaner code.

irog-c committed:
Update email splitter function and add tests

KyryloKupin committed:
Refactor EmailVerification and update tests

This commit refactors the EmailVerification.cpp, rearranging include orders and declaring a variable within an if condition for clean and efficient code. Additionally, the structure of the test file has been slightly modified to improve readability and maintainability.

irog-c committed:
Add function implementation for checking mx record + clang format fix

irog-c committed:
Change request

irog-c committed:
Tests

irog-c committed:
Check error code during socket connect

KyryloKupin committed:
Refactor email verification system for better testing capabilities

This commit modifies the email verification system by dividing some functionalities into separate components to facilitate better unit testing. This involves restructuring code, renaming classes, creating new socket wrapper files, extracting TCP socket related code to a socket wrapper, and updating associated tests accordingly. Some obsolete tests and configurations have also been removed from the system. Additional changes include switching to `find_package` for installing packages, enforcing standard conformance on MSVC, and updating google test version.

KyryloKupin committed:
Add function implementation for checking mx record + clang format fix

KyryloKupin committed:
Update CMakeLists and project settings

Refactored CMakeLists.txt files simplifying project setup. Introduced .clang-tidy and .clang-format for standardized code formatting and linting. Additionally, updated project name and version, and removed unused options related to test coverage.

KyryloKupin committed:
Change build type to Release and remove VRF test

The build configuration in 'ubuntu.yml' was changed from Debug to Release for better optimization and speed. An unnecessary VRF test, 'CHECK_MX_1', was also removed from 'vrf.cpp' to streamline the testing process. The test command in the 'ubuntu.yml' file was updated to include extra verbosity for better logs.

KyryloKupin committed:
Update build command in Github Actions workflow

The build command in the Ubuntu Github Actions workflow has been updated. Instead of hardcoding the number of parallel jobs to 4, the build command now dynamically gets the number of logical CPUs available on the Github Actions runner using the 'sysctl' command. This allows for faster build times on runners with more than 4 CPUs.

KyryloKupin committed:
Update build step in Github workflow

The build step in the Github Actions workflow for Ubuntu has been updated. The change involves removing the specification of the number of logical CPUs used in parallel builds. This simplifies the command and could possibly mitigate potential issues related to resource allocation.

irog-c committed:
Add tuporesolv dependency to project

KyryloKupin committed:
Remove c-ares

KyryloKupin committed:
Remove asio with boost_asio

KyryloKupin committed:
Add new verifier class and refactor vrf related data and functions

KyryloKupin committed:
Reformat

KyryloKupin committed:
Rename namespace and add standalone project setup

Renamed namespace from `tuposoft::vrf` to `tuposoft::tupovrf` throughout the codebase. Added a CMakeLists.txt for setting up standalone integration, including Boost dependencies and a new executable for the verifier.

KyryloKupin committed:
Run the IO on the requested number of threads

KyryloKupin committed:
Add cxxopts lib to handle CLI arguments

KyryloKupin committed:
Replace tuporesolv with asio_dns_resolver
